### PR TITLE
Add cache support for the capabilites request #734

### PR DESCRIPTION
**Pull request by prb112**
- Update cloud.xml to point to the location ( per warning in eclipse)
- Update Capabilities statement to use a concurrent hash map
- Cache time is one hour

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

**Comment by prb112** [comment]
Added John as he commented on the issue

File: fhir-server/liberty-config/configDropins/disabled/cloud.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <server>
     <!-- webApp security is MUST be provided via a gateway -->
```
Comment (lmsurpre) [inline fhir-server/liberty-config/configDropins/disabled/cloud.xml:2]: not related to your change at all, but i just spotted this typo that i must have introduced earlier:
```suggestion
    <!-- webApp security MUST be provided via a gateway -->
```

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java [hunk 2]
```diff
@@ -388,26 +404,23 @@ private CapabilityStatement addExtensionElements(CapabilityStatement capabilityS
...
+                        .build();
         extentions.add(extension);
 
-        extension = Extension.builder()
-                .url(EXTENSION_URL + "/persistenceType")
-                .value(string(getPersistenceImpl().getClass().getSimpleName()))
-                .build();
+        extension =
+                Extension.builder().url(EXTENSION_URL + "/persistenceType")
+                        .value(string(getPersistenceImpl().getClass().getSimpleName())).build();
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:423]: i prefer the style we had before

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java [hunk 3]
```diff
@@ -330,56 +346,56 @@ private CapabilityStatement buildCapabilityStatement() throws Exception {
...
 
-        extension = Extension.builder()
-                .url(EXTENSION_URL + "/natsNotificationsEnabled")
-                .value(com.ibm.fhir.model.type.Boolean.of(fhirConfig.getBooleanProperty(FHIRConfiguration.PROPERTY_NATS_ENABLED, Boolean.FALSE)))
-                .build();
+        extension =
+                Extension.builder().url(EXTENSION_URL + "/natsNotificationsEnabled")
+                        .value(com.ibm.fhir.model.type.Boolean.of(
+                                fhirConfig.getBooleanProperty(FHIRConfiguration.PROPERTY_NATS_ENABLED, Boolean.FALSE)))
+                        .build();
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:385]: I prefer the style we had before

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java [hunk 4]
```diff
@@ -285,28 +307,22 @@ private CapabilityStatement buildCapabilityStatement() throws Exception {
...
+                                Code.of(FHIRMediaType.APPLICATION_FHIR_XML))
+                        .version(string(buildInfo.getBuildVersion())).name(string(FHIR_SERVER_NAME))
+                        .description(Markdown.of(buildDescription)).copyright(Markdown.of(FHIR_COPYRIGHT))
+                        .publisher(string("IBM Corporation"))
+                        .software(CapabilityStatement.Software.builder().name(string(FHIR_SERVER_NAME))
+                                .version(string(buildInfo.getBuildVersion())).id(buildInfo.getBuildId()).build())
+                        .rest(rest)
+                        .instantiates(
+                                Canonical.of("http://www.hl7.org/fhir/bulk-data/CapabilityStatement-bulk-data.html"))
+                        .build();
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:326]: I prefer the style we had before

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java [hunk 5]
```diff
@@ -245,38 +272,33 @@ private CapabilityStatement buildCapabilityStatement() throws Exception {
...
+                                .build())
+                                .text(string("OAuth2 using SMART-on-FHIR profile (see http://docs.smarthealthit.org)"))
+                                .build())
+                        .extension(Extension.builder()
+                                .url("http://fhir-registry.smarthealthit.org/StructureDefinition/oauth-uris")
+                                .extension(Extension.builder().url("token").value(Url.of(tokenURL)).build(),
+                                        Extension.builder().url("authorize").value(Url.of(authURL)).build(),
+                                        Extension.builder().url("register").value(Url.of(regURL)).build())
+                                .build())
+                        .build();
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:288]: i prefer the style we had before

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java [hunk 6]
```diff
@@ -200,19 +230,17 @@ private CapabilityStatement buildCapabilityStatement() throws Exception {
...
+                    Rest.Resource.builder().type(ResourceType.of(resourceType))
+                            .profile(Canonical.of("http://hl7.org/fhir/profiles/" + resourceTypeName))
+                            .supportedProfile(FHIRRegistry.getInstance().getProfiles(resourceTypeName))
+                            .interaction(interactions).operation(ops)
+                            .conditionalCreate(com.ibm.fhir.model.type.Boolean.of(true))
+                            .conditionalUpdate(com.ibm.fhir.model.type.Boolean.of(true))
+                            .updateCreate(com.ibm.fhir.model.type.Boolean.of(isUpdateCreateEnabled()))
+                            .conditionalDelete(ConditionalDeleteStatus.MULTIPLE)
+                            .conditionalRead(ConditionalReadStatus.FULL_SUPPORT).searchParam(conformanceSearchParams)
+                            .build();
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:243]: i prefer the style we had before

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java [hunk 7]
```diff
@@ -178,15 +210,13 @@ private CapabilityStatement buildCapabilityStatement() throws Exception {
                 for (SearchParameter searchParameter : searchParameters) {
                     // The name here is a natural language name, and intentionally not replaced with code.
                     Rest.Resource.SearchParam.Builder conformanceSearchParamBuilder =
-                            Rest.Resource.SearchParam.builder()
-                                .name(searchParameter.getName())
-                                .type(searchParameter.getType());
+                            Rest.Resource.SearchParam.builder().name(searchParameter.getName())
+                                    .type(searchParameter.getType());
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:214]: i prefer the old style

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java [hunk 8]
```diff
@@ -71,54 +74,83 @@
...
+            ZonedDateTime cachedTime = ZonedDateTime.from(acc);
+
+            if (ZonedDateTime.now().isBefore(cachedTime.plusHours(1))) {
+                statement = getCapabilityStatement();
+            }
         }
+        return statement;
     }
 
     private synchronized CapabilityStatement getCapabilityStatement() throws FHIROperationException {
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:148]: I've always wondered why this method is synchronized.  Definitely orthogonal to your PR, but let me know if you agree it shouldn't be needed...

File: fhir-server/liberty-config/configDropins/disabled/cloud.xml (hunk 1 above)
Comment (prb112) [reply fhir-server/liberty-config/configDropins/disabled/cloud.xml:2]: Fixing in my upcoming pr

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java (hunk 8 above)
Comment (JohnTimm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:148]: I think you are going to want to use the compute method on ConcurrentHashMap

Comment (JohnTimm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:148]: and computeIfAbsent for the first one

Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:148]: Let me remove. If we find we need to add it back we can.

Comment (JohnTimm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:148]: what about something like:
```
CAPABILITY_STATEMENT_CACHE.compute(tenantId, k -> getOrCreateCapabilityStatement());
```

Comment (JohnTimm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:148]: compute should be atomic (for ConcurrentHashMap)

Comment (JohnTimm) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:148]: and compute returns the new value:
```
CapabilityStatement capabilityStatement = CAPABILITY_STATEMENT_CACHE.compute(tenantId, k -> getOrCreateCapabilityStatement(tenantId));
```

**Comment by prb112** [comment]
Flipped the code around a bit per a chat with John

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java (hunk 8 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java]: Flipped around per conversation

**Comment by prb112** [comment]
Initial request is 5 seconds, and subsequent requests take roughly 2 seconds (realtime calculations using `time` on curl)

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 9]
```diff
@@ -1443,6 +1443,7 @@ This section contains reference information about each of the configuration prop
 |`fhirServer/core/checkReferenceTypes`|boolean|Indicates whether reference type checking is performed by the server during parsing / deserialization.|
 |`fhirServer/core/serverRegistryResourceProviderEnabled`|boolean|Indicates whether the server registry resource provider should be used by the FHIR registry component to access definitional resources through the persistence layer.|
 |`fhirServer/core/conditionalDeleteMaxNumber`|integer|The max number of matches supported in conditional delete. |
+|`fhirServer/core/capabilityStatementCacheTimeout`|integer|The max time a CapabilityStatement is cached for the metadata endpoint. |
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1446]: The description needs to specify the units.
```suggestion
|`fhirServer/core/capabilityStatementCacheTimeout`|integer|The number of minutes that a tenant's CapabilityStatement is cached for the metadata endpoint. |
```

**Comment by lmsurpre** [comment]
I still wonder if we could add some HTTP cache-control headers to hint that clients and/or proxies should cache it on their side for the specified time-period.  However, since its tenant-specific, that could be error-prone if not done with care and this PR is definitely an improvement on the current state of affairs, so I'm good with it pending my 1 doc suggestion.

**Comment by prb112** [comment]
> I still wonder if we could add some HTTP cache-control headers to hint that clients and/or proxies should cache it on their side for the specified time-period. However, since its tenant-specific, that could be error-prone if not done with care and this PR is definitely an improvement on the current state of affairs, so I'm good with it pending my 1 doc suggestion.

Added a cache-control header and marked as Private, it considers the tenant timeout length. 
It's a rough implementation.

**Pull request merged**